components/sensor/ina219/ina219.go

```diff
@@ -0,0 +1,230 @@
+		bus:    i2cbus,
+		addr:   byte(addr),
+	}
+
+	err = s.calibrate(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	return s, nil
+}
+
+// ina219 is a i2c sensor device that reports voltage, current and power
+type ina219 struct {
+	generic.Unimplemented
+	logger     golog.Logger
+	bus        board.I2C
+	addr       byte
+	currentLSB int64
+	powerLSB   int64
+	cal        uint16
+}
+
+type PowerMonitor struct {
+	Shunt   int64
+	Voltage float64
+	Current float64
+	Power   float64
+}
+
+func (d *ina219) calibrate(ctx context.Context) error {
+	if senseResistor <= 0 {
+		return fmt.Errorf("ina219 calibrate: senseResistor value invalid %d", senseResistor)
+	}
+	if maxCurrent <= 0 {
+		return fmt.Errorf("ina219 calibrate: maxCurrent value invalid %d", maxCurrent)
+	}
+
+	d.currentLSB = maxCurrent / (1 << 15)
+	d.powerLSB = int64((maxCurrent*20 + (1 << 14)) / (1 << 15))
+	// Calibration Register = 0.04096 / (current LSB * Shunt Resistance)
+	// Where lsb is in Amps and resistance is in ohms.
+	// Calibration register is 16 bits.
+	cal := calibratescale / (d.currentLSB * int64(senseResistor))
+	if cal >= (1 << 16) {
+		return fmt.Errorf("ina219 calibrate: calibration register value invalid %d", cal)
+	}
+	d.cal = uint16(cal)
+
+	return nil
+}
+
+// Readings returns a list containing three items (voltage, current, and power).
+func (d *ina219) Readings(ctx context.Context, extra map[string]interface{}) (map[string]interface{}, error) {
+	handle, err := d.bus.OpenHandle(d.addr)
+	if err != nil {
+		d.logger.Errorf("can't open ina219 i2c: %s", err)
+		return nil, err
+	}
+	defer handle.Close()
+
+	// use the calibration result to set the scaling factor
+	// of the current and power registers for the maximum resolution
+	buf := make([]byte, 2)
+	binary.BigEndian.PutUint16(buf, d.cal)
+	err = handle.WriteBlockData(ctx, calibrationRegister, buf)
+	if err != nil {
+		return nil, err
+	}
+
+	buf = make([]byte, 2)
+	binary.BigEndian.PutUint16(buf, uint16(0x1FFF))
+	err = handle.WriteBlockData(ctx, configRegister, buf)
+	if err != nil {
+		return nil, err
+	}
+
+	var pm PowerMonitor
+
+	// get shunt voltage - currently we are not returning - is it useful?
+	shunt, err := handle.ReadBlockData(ctx, shuntVoltageRegister, 2)
+	if err != nil {
+		return nil, err
+	}
+
+	// Least significant bit is 10µV.
+	pm.Shunt = int64(binary.BigEndian.Uint16(shunt)) * 10 * 1000
+
+	bus, err := handle.ReadBlockData(ctx, busVoltageRegister, 2)
+	if err != nil {
+		return nil, err
+	}
+
+	// Check if bit zero is set, if set the ADC has overflowed.
+	if binary.BigEndian.Uint16(bus)&1 > 0 {
+		return nil, fmt.Errorf("ina219 bus voltage register overflow")
+	}
+
+	pm.Voltage = float64(binary.BigEndian.Uint16(bus)>>3) * 4 / 1000
+
+	current, err := handle.ReadBlockData(ctx, currentRegister, 2)
+	if err != nil {
+		return nil, err
+	}
+
+	pm.Current = float64(int64(binary.BigEndian.Uint16(current))*d.currentLSB) / 1000000000
+
+	power, err := handle.ReadBlockData(ctx, powerRegister, 2)
```

Comment:
The registers you're reading from are all contiguous. On most I2C devices, if you want to read contiguous registers, you can just read off the end of one and into the start of the next. Does that technique work with this sensor? If so, we can do that to nearly double the speed of this function:
```go
allRawData := handle.ReadBlockData(ctx, shuntVoltageRegister, 8)
shunt   := binary.BigEndian.Uint16(allRawData[0:2])
bus     := binary.BigEndian.Uint16(allRawData[2:4])
power   := binary.BigEndian.Uint16(allRawData[4:6])
voltage := binary.BigEndian.Uint16(allRawData[6:8])
```
Each time you start a new `ReadBlockData`, you send 2 bytes (the device address and the register to start at), so this approach nearly halves the amount of communication you need to do (from 16 bytes down to 10, being 8 bytes of data read in plus 2 for each communication).